### DISCUSSION TITLE
client.search

### DISCUSSION THREAD

**Discussion by dennisbrox** [FirelyTeam/firely-net-sdk#1964] [Q&A] [answered]
Are there any examples anywhere of how you search in an identifier item for value, use, type.text, etc.  I don't want to use the Id field that sdk provides.

**Comment by mbaltus** [comment]
The Id and Identifier fields are two different things.
`Id` holds the technical identity of a resource, which has a simple string value. It is used for retrieving the resource from a system and for linking to the resource.
`Identifier` holds business/administrative identifiers for a resource and is of type Identifier, which has the attributes you mention, like the value, system, use, etc.

Can you clarify what you mean by 'search in an identifier'? Is that a search you want to send to a server, or do you want to look at those fields from within your code?

**Comment by dennisbrox** [comment]
"identifier" is the name of the property in the FHIR specification.
"search on an identifier" means what it says and does not permit B-Tree
searching like _id does.  Incidentally, Screwing up the interface between a
language (Identifier) and the FHIR spec (identifier) because of
capitalization is just such  technical nonsense it's hard to believe.

**Reply by mbaltus** [reply]
Thanks for your input.

So you want to perform a search on a FHIR server using the fields in the identifier?
Searching a resource by identifier.system and identifier.value is the default, see the description on token type search parameters: http://hl7.org/fhir/search.html#token.
You can also search on the identifier.type field, using the 'of-type' modifier described in that paragraph, if the server has implemented that modifier.
The syntax for a token type search can be used in the SearchParams object in the SDK, so you can do a FhirClient.Search with that.

**Comment by dennisbrox** [comment]
No, I want to do a search using _id which is always in a B-Tree .

**Comment by mbaltus** [comment, marked as answer]
You can do a search on _id, with FhirClient.SearchById or using a SearchParams with "_id=[value]" and the FhirClient.Search method.
However, you asked about searching on an identifier and its use, type, etc. fields, which is why I asked for more information.

A FHIR resource's id is a technical thing that just has a value. It is meant for easy, quick lookups. For example it could be the primary key in a database table. It's not an identifier. It does not have any associated attributes like type or use. So while you can search on that value, there is no notion of a 'use', 'type', 'system', etc. and you cannot do a search on those using the _id search parameter.

**Comment by dennisbrox** [comment]
I know.  Problem is I can't set _id in a resource update PUT to the AWS
FHIR works server when the resource doesn't exist and "technical support"
there does not  know how to configure that in their Lambda service.

**Reply by mbaltus** [reply]
I see. That's unfortunately something we cannot change for you or help you with. Maybe you can add a ticket on one of the AWS FHIR Works github repos to indicate you want to use 'Update as Create' giving you control over the technical ids.
